docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,197 @@
+
+## Changes
+
+### New APIs to query and select registered attesters and verifiers
+
+```C
+/**
+ * oe_get_registered_attester_format_ids
+ *
+ * Get the format IDs of all registered attesters.
+ *
+ * @param[out] format_ids The list of the format ids of
+ * the registered attesters.
+ * @param[out] format_ids_length The length of the format ids list.
+ * @retval OE_OK on success.
+ * @retval Otherwise, returns the error code the plugin's function.
+ */
+oe_result_t oe_get_registered_attester_format_ids(
+    oe_uuid_t** format_ids,
+    size_t* format_ids_length);
+
+/**
+ * oe_get_registered_verifier_format_ids
+ *
+ * Get the format IDs of all registered verifiers.
+ *
+ * @param[out] format_ids The list of the format ids of
+ * the registered verifiers.
+ * @param[out] format_ids_length The length of the format ids list.
+ * @retval OE_OK on success.
+ * @retval Otherwise, returns the error code the plugin's function.
+ */
+oe_result_t oe_get_registered_verifier_format_ids(
+    oe_uuid_t** format_ids,
+    size_t* format_ids_length);
+
+/**
+ * oe_free_format_ids
+ *
+ * Frees the format ids.
+ *
+ * @param[in] format_ids The list of the format ids.
+ * @retval OE_OK on success.
+ */
+oe_result_t oe_free_format_ids(oe_uuid_t* format_ids);
+
+/**
+ * oe_select_attester_format
+ *
+ * From the list, select an evidence format that the registered attesters
+ * can generate.
+ *
+ * @param[in] format_ids The list of the format ids.
+ * @param[in] format_ids_length The length of the format ids list.
+ * @param[out] selected_format_id The selected id from the evidence
+ * format ids list.
+ * @retval OE_OK on success.
+ * @retval Otherwise, returns the error code the plugin's function.
+ */
+oe_result_t oe_select_attester_format(
+    const oe_uuid_t* format_ids,
+    size_t format_ids_length,
+    oe_uuid_t** selected_format_id);
+```
+### Built-in SGX plugin splitting
+```C
+#if (OE_API_VERSION < 3)
+/**
+ * Helper function that returns the SGX attester that can then be sent to
+ * `oe_register_attester`.
+ * 
+ * @deprecated This function is deprecated. Use oe_sgx_plugin_local_attester
+ * or oe_sgx_plugin_dcap_attester instead.
+ *
+ * @retval A pointer to the SGX attester. This function never fails.
+ */
+oe_attester_t* oe_sgx_plugin_attester(void);
+#endif
+
+/**
+ * Helper function that returns the SGX local attester that can then be sent to
+ * `oe_register_attester`.
+ *
+ * @retval A pointer to the SGX attester. This function never fails.
+ */
+oe_attester_t* oe_sgx_plugin_local_attester(void);
+
+/**
+ * Helper function that returns the SGX DCAP (ECDSA P256) attester that can then
+ * be sent to `oe_register_attester`.
+ *
+ * @retval A pointer to the SGX attester. This function never fails.
+ */
+oe_attester_t* oe_sgx_plugin_dcap_attester(void);
```

Comment:
Is the plan of splitting up the SGX plugins based off local/DCAP/ASEM or by algorithm? Will there be multiple plugins in the future to support different intel quote formats like EPID?

Reply:
If the way to generate or verify the evidence is unique, it is a new plugin. Multiple plugins are planned to support different Intel quote formats.

---
In reply to: [386736293](https://github.com/openenclave/openenclave/pull/2621#discussion_r386736293) [](ancestors = 386736293)

Reply:
I think the best approach is upon enclave load / initialization, the enclave plugin library enumerates supported format IDs and self-register plugins for those IDs. With this approach, application software don't have to be exposed the internals of oe_attester_t structures. They simply need to call oe_get_registered_attester_format_ids() to discover which IDs are supported / registered, or to call oe_select_attester_format() to select a supported from the input list.

Alternatively upon initialization, the enclave plugin library could enumerate all supported format IDs, and creates a list of attester plugins (list of oe_attester_t structures). OE SDK could have an API, such as "oe_result_t oe_sgx_get_attester_plugins(oe_attester_t** attesters, size_t attesters_length)", to retrieve the list, and register them one by one with oe_register_attester().

With this approach, in an SGX enclave, the list returned by oe_sgx_get_attester_plugins() would always include a plugin for OE_SGX_LOCAL_ATTESTATION_UUID. If DCAP library is used, a plugin for OE_SGX_ECDSA_P256_ATTESTATION_UUID would be included. If AESM-based library is used (instead of DCAP) that supports ECDSA and EPID, plugins for OE_SGX_ECDSA_P256_ATTESTATION_UUID, OE_SGX_EPID_LINKSABLE_ATTESTATION_UUID and OE_SGX_EPID_UNLINKSABLE_ATTESTATION_UUID would be included in the list.

Reply:
Both options require the application to specify force-load flag when linking to the plugin libraries (so they can do their initialization sequences and register themselves). The application developer may miss the linking flag and wonder why nothing is registered. Also, the app developer may not realize the plugins implicitly register themselves which creates possible security vulnerabilities.

OE SDK does not prevent plugins from registering themselves at initialization. In such case, if the application register them again, ALREADY_REGISTERED is returned and the app can choose to ignore the error. Ultimately, the application should be aware of which plugins it is carrying with.

Reply:
In the current OE SDK configuration, all the enclave-side attestation code is built into the single liborcore.a library, and the host-side code is included in the liboehost.a library. So does not seem there is a need of force-load flag.

In any case, I think it would be helpful to clarify which of the possible protocols are required to be supported by all TEEs, so that application software that implements this required protocol is guaranteed to work in all TEEs, thus being TEE agnostic.

Reply:
To avoid the complication of initialization-time plugin registration and to maintain TEE agnostic, there are several options. Among them:

Option 1. The SDK exposes two TEE agnostic initialization functions "oe_result_t oe_sgx_get_attester_plugins(oe_attester_t** attesters, size_t attesters_length)" and "oe_result_t oe_sgx_get_verifier_plugins(oe_verifier_t** verifiers, size_t verifiers_length)", as describe in a previous comment. These functions are exposed instead of the SGX specific ones oe_sgx_plugin_local_attester() and oe_sgx_plugin_dcap_attester().

Application software’s call to one of these functions triggers the underlying plugin library to initialize itself and enumerate supported plugins. Application software can then register the returned list of plugins.

Option 2: The SDK exposes a single TEE agnostic initialization function “oe_result_t oe_initiailze_plugins()”.

Application software call to this function triggers the underlying plugin library to initialize itself, enumerate supported plugins and register all of them.

This option is better, since TEE-agnostic application software is not supposed to have to handle the plugins that are TEE and platform specific.

There are several approaches to maintain TEE agnostic. Among them:

Approach 1: in SDK headers, these TEE agnostic functions are macro’ed into TEE specific functions that are implemented by TEE specific plugin libraries.

This approach avoids the need of force-load linker flags.

Approach 2: in run-time, when a plugin library is loaded / initialized, it simply registers its initialization function with the SDK framework. Later when application software invokes a TEE agnostic initialization function, all registered library initialization functions are invoked, for plugin enumeration and registration.

In both approaches, the SDK framework and plugin library implementation need to ensure that application software can invoke the TEE-agnostic initialization functions multiple times (idempotent API?), to allow dynamic load / unload of plugin libraries.

Reply:
Offline discussion - the application has to explicitly register attesters and verifiers.

---
In reply to: [387267067](https://github.com/openenclave/openenclave/pull/2621#discussion_r387267067) [](ancestors = 387267067)

Reply:
Offline discussion - concluded on the alternative suggestion above. The enclave plugin library enumerates all supported format IDs, and creates a list of attester plugins.

---
In reply to: [387224634](https://github.com/openenclave/openenclave/pull/2621#discussion_r387224634) [](ancestors = 387224634)

Reply:
Offline discussion - the enclave plugin library enumerates all supported format IDs, and creates a list of attester plugins.

---
In reply to: [387353126](https://github.com/openenclave/openenclave/pull/2621#discussion_r387353126) [](ancestors = 387353126)

Reply:
Offline discussion - concluded on option 1.

---
In reply to: [389868506](https://github.com/openenclave/openenclave/pull/2621#discussion_r389868506) [](ancestors = 389868506)